generals/remote/generalsio_client.py

```diff
@@ -0,0 +1,109 @@
+from socketio import SimpleClient  # type: ignore
+
+from generals.agents.agent import Agent
+
+
+class GeneralsBotError(Exception):
+    """Base generals-bot exception
+    TODO: find a place for exceptions
+    """
+
+    pass
+
+
+class GeneralsIOClientError(GeneralsBotError):
+    """Base GeneralsIOClient exception"""
+
+    pass
+
+
+class RegisterAgentError(GeneralsIOClientError):
+    """Registering bot error"""
+
+    def __init__(self, msg: str) -> None:
+        super().__init__()
+        self.msg = msg
+
+    def __str__(self) -> str:
+        return f"Failed to register the agent. Error: {self.msg}"
+
+
+class GeneralsIOClient(SimpleClient):
+    """
+    Wrapper around socket.io client to enable Agent to join
+    GeneralsIO lobby.
+    """
+
+    def __init__(self, agent: Agent, user_id: str):
+        super().__init__()
+        self.connect("https://botws.generals.io")
+        self.user_id = user_id
+        self._queue_id = ""
+
+    @property
+    def queue_id(self):
+        if not self._queue_id:
+            raise GeneralsIOClientError("Queue ID is not set.\nIs agent in the game lobby?")
+
+        return self._queue_id
+
+    def _emit_receive(self, *args):
+        self.emit(*args)
+        return self.receive()
+
+    def register_agent(self, username: str) -> None:
+        """
+        Register Agent to GeneralsIO platform.
+        You can configure one Agent per `user_id`. `user_id` should be handled as secret.
+        :param user_id: secret ID of Agent
+        :param username: agent username, must be prefixed with `[Bot]`
+        """
+        event, response = self._emit_receive("set_username", (self.user_id, username))
+        if response:
+            # in case of success the response is empty
+            raise RegisterAgentError(response)
+
+    def join_private_game(self, queue_id: str) -> None:
+        """
+        Join (or create) private game lobby.
+        :param queue_id: Either URL or lobby ID number
+        """
+        self._emit_receive("join_private", (queue_id, self.user_id))
+        self._queue_id = queue_id
+
+    def wait_for_game(self, force_start: bool = True) -> None:
+        """
+        Wait for the game start.
+        :param force_start: If set to True, the Agent will set `Force Start` flag
+        """
+        if force_start:
+            self.emit("set_force_start", (self.queue_id, True))
+
+        while True:
+            event = self.receive()[0]
+            if event == "game_start":
+                break
+
+        self._start_game()
+
+    def _start_game(self) -> None:
+        """
+        Triggered after server starts the game.
+        TODO: spawn a new thread in which Agent will calculate its moves
+        """
+        winner = False
+        while True:
```

Comment:
same question here, maybe it is good to look up best practices (i can look it up)